src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,191 @@
+        assert sources_field.raw_value == raw_source_files
+        result: HaskellSourcesResult = run_rule(
+            hydrate_haskell_sources,
+            rule_args=[sources_field],
+            mock_gets=[
+                MockGet(
+                    product_type=Snapshot,
+                    subject_type=PathGlobs,
+                    mock=lambda _: Snapshot(
+                        directory_digest=EMPTY_DIRECTORY_DIGEST,
+                        files=hydrated_source_files,
+                        dirs=(),
+                    ),
+                )
+            ],
+        )
+        return result
+
+    # Normal field
+    expected_files = ("monad.hs", "abstract_art.hs", "abstract_algebra.hs")
+    assert (
+        hydrate_field(
+            raw_source_files=["monad.hs", "abstract_*.hs"], hydrated_source_files=expected_files
+        ).snapshot.files
+        == expected_files
+    )
+
+    # Test pre-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=[0, 1, 2], hydrated_source_files=())  # type: ignore[call-arg]
+    assert "Not all elements of the iterable have type" in str(exc)
+
+    # Test post-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=["*.js"], hydrated_source_files=("not_haskell.js",))
+    assert "Received non-Haskell sources" in str(exc)
+
+
+def test_has_fields() -> None:
+    class UnrelatedField(PrimitiveField):
+        alias: ClassVar = "unrelated"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    tgt = HaskellTarget({})
+    assert tgt.has_fields([]) is True
+    assert tgt.has_fields([HaskellGhcExtensions]) is True
+    assert tgt.has_fields([UnrelatedField]) is False
+    assert tgt.has_fields([HaskellGhcExtensions, UnrelatedField]) is False
+
+
+def test_field_hydration_is_lazy() -> None:
+    bad_extension = "DoesNotStartWithGhc"
+    # No error upon creating the Target because validation does not happen until a call site
+    # hydrates the specific field.
+    tgt = HaskellTarget(
+        {HaskellGhcExtensions.alias: ["GhcExistentialQuantification", bad_extension]}
+    )
+    # When hydrating, we expect a failure.
+    with pytest.raises(ValueError) as exc:
+        tgt.get(HaskellGhcExtensions).value
+    assert "must be prefixed by `Ghc`" in str(exc)
+
+
+def test_add_custom_fields() -> None:
+    class CustomField(PrimitiveField):
+        alias: ClassVar = "custom_field"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    union_membership = UnionMembership(OrderedDict({HaskellField: OrderedSet([CustomField])}))
```

Comment:
Is the idea here that every Target type will have its own union? There are no union parent-child relationships, so will we end up with separate `PythonBinaryField` and `PythonLibraryField` and `PythonTestsField` unions which are 90% the same but each has a few extra elements?

Reply:
Yes, although, those unions are only for plugin fields, meaning users are adding additional custom fields to pre-existing target types.

So, a user's `register.py` would likely look like this:

```python
def rules():
  return [UnionRule(python_tgt, TypeChecked) for python_tgt in [PythonBinary, PythonLibrary, Pythontests])
```